### ISSUE TITLE
Proposal: have a waitForFlag for suspend/resume

### ISSUE THREAD

**Issue by jstrachan** [argoproj/argo#702]
It would be great to allow a workflow to suspend itself until some named flag is set on the workflow instance. Eg if a workflow needs to wait for an external event (a helm install completed, a PR merged, an issue was closed etc)

From an implementation perspective the ‘waitForFlag’ could just suspend a workflow, recording its current step. Then separately an external event handler, like Prow, could be used to handle incoming GitHub events (eg a PR merged or issue closed) and update the named flag on the workflow instance - which the Argo controller could detect and then resume the workflow.

**Comment by jessesuen** [comment]
@jstrachan I agree that the ability to stop at a predetermined step in the workflow and resume asynchronously sometime later would be very useful feature. We had something similar to this in argo v1 called an "approval" container which no longer exists in argo v2. In Argo v1 the workflow would run a container which was implemented as a dumb REST server, which waited for a special HTTP request to resume or fail. I think we can do better than that, with first class support for pause and resume in the workflow controller.

Based on your requirements, what would be your thoughts about a new "breakpoint" template type specially for this purpose? The benefit of introducing a new template type would allow us to introduce more control knobs of how we wanted to suspend and resume vs. a single binary waitForFlag. For example after suspending the workflow, we could have a timeout decision would fail itself if no one ever resumed it after some period of time.

I should also add that what you are asking for should already be possible (in a clunky way) by implementing something like a dumb sleep container which waits for a signal from an external entity. The event handler could do something simple like `kubectl exec <pod> sh -c kill -SIGUSR1 1` to indicate that it should exit with code 0 and proceed with the workflow.

**Comment by jstrachan** [comment]
@jessesuen that all sounds great to me. Having a generic pause / resume or breakpoint step would do the trick. 

I like the idea of keeping the logic inside the workflow (either the dsl or inside the step/container) but having a way to wake up a paused workflow from a custom event handler / web hook.

It might be nice to have a ‘wait until

**Comment by jstrachan** [comment]
whoops I didn’t mean to close the issue -  my phone screen isn’t quite big enough ;)

**Comment by jstrachan** [comment]
the general idea is when waiting for human approval or long system / soak tests to complete, it would be awesome to pause a workflow so it uses no runtime pods until a timer fires or an event handler / UI / CLI triggers a resume

**Comment by jessesuen** [comment]
> Then when there is no external event forcing an early resume, you can still implement waiting on long term events without keeping pods around - just polling with short term checks.

I agree that pods should not be kept around for a paused workflow -- in no way am I suggesting to have a special container handle the resume via some external HTTP/signal. That's one thing I never liked about the approval container technique. I think the right approach would be the suggestion you already made earlier by updating the workflow resource itself. The act of updating the workflow would automatically trigger an update event in the controller and requeue the workflow for the controller to re-evaluate the next steps.

As far as implementation, in my mind, resuming a workflow would translate to an HTTP PATCH / PUT to the k8s api server against the workflow resource, by update the specific paused node (step) in the graph, to either a Failed or Succeeded state. The advantage of leveraging the k8s api server is we get RBAC for free.

**Comment by jstrachan** [comment]
Being able to add breakpoints to workflows would be really useful for debugging too ;) For a debugger having a ‘resumeOneStepThenPause’ operation would be super handy to let you step through a paused workflow from a breakpoint. Though just having a pause step & resume action would be good enough to start with ;)

**Comment by jessesuen** [comment]
Looking to have this as part of v2.1 release. A new template type, `suspend`, will be introduced. And referenced in a steps/DAG template as follows:

```
templates:
- name: wait-for-approval
  suspend: {}

- name: release
  steps:
  - - name: build
      template: build
  - - name: wait-for-approval
      template: wait-for-approval
  - - name: release
      template: release
```

When execution reaches that point in the workflow, the workflow will become suspended, but only in that branch of the tree. Other branches of the workflow tree will continue to run if they are not gated by a suspend node. A new command will be introduced: `argo resume`. `argo resume` would resume an entire workflow (all nodes that are suspended). But it should also have the ability to resume a single suspended node in the workflow tree (if multiple branches of the tree are suspended).

**Comment by jstrachan** [comment]
that sounds great thanks!
